src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs

```diff
@@ -88,25 +88,22 @@ public abstract class RegexRunner
         protected Match? Scan(Regex regex, string text, int textbeg, int textend, int textstart, int prevlen, bool quick) =>
             Scan(regex, text, textbeg, textend, textstart, prevlen, quick, regex.MatchTimeout);
 
-        protected internal Match? Scan(Regex regex, string text, int textbeg, int textend, int textstart, int prevlen, bool quick, TimeSpan timeout)
+        protected internal virtual void Scan(ReadOnlySpan<char> text)
         {
-            this.quick = quick;
-
-            // Handle timeout argument
-            _timeout = -1; // (int)Regex.InfiniteMatchTimeout.TotalMilliseconds
-            bool ignoreTimeout = _ignoreTimeout = Regex.InfiniteMatchTimeout == timeout;
-            if (!ignoreTimeout)
+            string? s = runtext;
+            if (text != s)
             {
-                // We are using Environment.TickCount and not Stopwatch for performance reasons.
-                // Environment.TickCount is an int that cycles. We intentionally let timeoutOccursAt
-                // overflow it will still stay ahead of Environment.TickCount for comparisons made
-                // in DoCheckTimeout().
-                Regex.ValidateMatchTimeout(timeout); // validate timeout as this could be called from user code due to being protected
-                _timeout = (int)(timeout.TotalMilliseconds + 0.5); // Round;
-                _timeoutOccursAt = Environment.TickCount + _timeout;
-                _timeoutChecksToSkip = TimeoutCheckFrequency;
+                s = text.ToString();
+                runtext = s;
```

Comment:
Ok, so you have this logic here, it just seems like the `runtext = input` logic for string inputs was missing earlier.

As noted, we could alternatively choose to throw a NotSupportedException here if text != s.  We should think through the implications.  On the one hand, it's nice to implicitly support the span-based APIs when using one of those older CompileToAssembly implementations.  On the other hand, this could result in a massive allocation increase when someone switches to using the span-based APIs, and it could easily be a pit of failure; an exception would immediately highlight the problem and help people to know not to use that combination (which should be a rare combination, I hope).